GocadRelatedScripts/Face.py

```diff
@@ -0,0 +1,195 @@
+
+    def setup_proj_objects(self, sProj):
+        import pyproj
+
+        lla = pyproj.Proj(proj="latlong", ellps="WGS84", datum="WGS84")
+        if sProj != "geocent":
+            myproj = pyproj.Proj(sProj)
+        else:
+            myproj = pyproj.Proj(proj="geocent", ellps="WGS84", datum="WGS84")
+        return lla, myproj
+
+    def proj(self, sProj):
+        "project the node coordinate array"
+        import pyproj
+
+        lla, myproj = self.setup_proj_objects(sProj)
+        print("projecting the node coordinates")
+        self.vertex[:, 0], self.vertex[:, 1], self.vertex[:, 2] = pyproj.transform(lla, myproj, self.vertex[:, 0], self.vertex[:, 1], self.vertex[:, 2], radians=False)
+        print(self.vertex)
+        print("done projecting")
+
+    def proj_to_latlon(self, sProj):
+        "project the node coordinate array"
+        import pyproj
+
+        lla, myproj = self.setup_proj_objects(sProj)
+        print("projecting the node coordinates to geographic coords")
+        self.vertex[:, 0], self.vertex[:, 1], self.vertex[:, 2] = pyproj.transform(lla, myproj, self.vertex[:, 0], self.vertex[:, 1], self.vertex[:, 2], radians=False)
+        print(self.vertex)
+        print("done projecting")
+
+    def scale_vertex(self, scale):
+        "convert vertex array to km"
+        for i in range(3):
+            self.vertex[:, i] = self.vertex[:, i] * scale[i]
+
+    def translate_vertex(self, translation):
+        "convert vertex array to km"
+        for i in range(3):
+            self.vertex[:, i] = self.vertex[:, i] + translation[i]
+
+    def compute_id_vertex(self):
+        "id_vertex is an array with the id of all vertex forming the face"
+        self.id_vertex = np.unique(self.connect.flatten()) + 1
+
+    def intersect(self, Face2):
+        self.compute_id_vertex()
+        Face2.compute_id_vertex()
+        "return the common nodes between self and Face2"
+        return np.intersect1d(self.id_vertex, Face2.id_vertex)
+
+    def reindex(self, vid_lookup):
+        print("reindexing triangles...")
+        for itr in range(self.ntriangles):
+            for iv in range(3):
+                self.connect[itr, iv] = vid_lookup[self.connect[itr, iv]]
+
+    def __writeTs(self, fname, vertex, write_full_vertex_array=True, append=False):
+        """ output face as a *.ts file
+             vertex: nodes coordinates array
+        """
+        if write_full_vertex_array:
+            vertex_id_2_write = range(1, len(vertex) + 1)
+        else:
+            self.compute_id_vertex()
+            vertex_id_2_write = self.id_vertex
+
+        mode = "a" if append else "w"
+        with open(fname, mode) as fout:
+            fout.write("GOCAD TSURF 1\nHEADER {\nname:%s\nborder: true\nmesh: false\n*border*bstone: true\n}\nTFACE\n" % (fname))
+            for ivx in vertex_id_2_write:
+                fout.write("VRTX %s %s %s %s\n" % (ivx, vertex[ivx - 1, 0], vertex[ivx - 1, 1], vertex[ivx - 1, 2]))
```

Comment:
Why do you have the -1 here and add a +1 in l132. IMHO could be just left out.

Reply:
If I remember well, Gocad does not like when the vertices in the ts files are indexed from 0.
It should start from 1.